calico-vpp-agent/policy/policy_server.go

```diff
@@ -1098,21 +1103,181 @@ func (s *Server) handleHostMetadataRemove(msg *proto.HostMetadataRemove, pending
 }
 
 func (s *Server) handleIpamPoolUpdate(msg *proto.IPAMPoolUpdate, pending bool) (err error) {
-	common.SendEvent(common.CalicoVppEvent{
-		Type: common.IpamPoolUpdate,
-		New:  msg,
-	})
+	if msg != nil {
+		err = s.handleIPPoolUpdate(msg, nil)
+		if err != nil {
+			return errors.Wrap(err, "error processing pool add / modified")
+		}
+	}
 	return nil
 }
 
 func (s *Server) handleIpamPoolRemove(msg *proto.IPAMPoolRemove, pending bool) (err error) {
+	if msg != nil {
+		err = s.handleIPPoolUpdate(nil, msg)
+		if err != nil {
+			return errors.Wrap(err, "error processing pool del")
+		}
+	}
+	return nil
+}
+
+// update updates the internal map with IPAM updates when the update
+// is new addtion to the map or changes the existing item, it calls
+// ipamUpdateHandler
+func (s *Server) handleIPPoolUpdate(poolUpdate *proto.IPAMPoolUpdate, poolRemove *proto.IPAMPoolRemove) error {
+	s.ippoolLock.Lock()
+	defer s.ippoolLock.Unlock()
+	key := ""
+	if poolUpdate != nil {
+		key = poolUpdate.Id
+	} else {
+		key = poolRemove.Id
+	}
+
+	if key == "" {
+		s.log.Debugf("Empty pool")
+		return nil
+	}
+
+	existing, found := s.ippoolmap[key]
+	if poolRemove != nil {
+		if found {
+			delete(s.ippoolmap, key)
+			s.log.Infof("Deleting pool: %s", key)
+			return s.ipamUpdateHandler(nil, &existing)
+		} else {
+			s.log.Warnf("Deleting unknown ippool")
+			return nil
+		}
+	} else {
+		if found && equalPools(poolUpdate, &existing) {
+			s.log.Infof("Unchanged pool: %s, nat:%t", key, poolUpdate.Pool.Masquerade)
+			return nil
+		} else if found {
+			s.log.Infof("Updating pool: %s, nat:%t", key, poolUpdate.Pool.Masquerade)
+			s.ippoolmap[key] = *poolUpdate
+
+			return s.ipamUpdateHandler(poolUpdate, &existing)
+		} else {
+			s.log.Infof("Adding pool: %s, nat:%t", key, poolUpdate.Pool.Masquerade)
+			s.ippoolmap[key] = *poolUpdate
+			return s.ipamUpdateHandler(poolUpdate, nil /* prevPool */)
+		}
+	}
+}
+
+// Compare only the fields that make a difference for this agent i.e. the fields that have an impact on routing
+func equalPools(a *proto.IPAMPoolUpdate, b *proto.IPAMPoolUpdate) bool {
+	if a.Pool.Cidr != b.Pool.Cidr {
+		return false
+	}
+	if a.Pool.IpipMode != b.Pool.IpipMode {
+		return false
+	}
+	if a.Pool.VxlanMode != b.Pool.VxlanMode {
+		return false
+	}
+	return true
+}
+
+func (s *Server) addDelSnatPrefix(pool *proto.IPAMPoolUpdate, isAdd bool) (err error) {
+	_, ipNet, err := net.ParseCIDR(pool.Pool.Cidr)
+	if err != nil {
+		return errors.Wrapf(err, "Couldn't parse pool CIDR %s", pool.Pool.Cidr)
+	}
+	if pool.Pool.Masquerade {
+		err = s.vpp.CnatAddDelSnatPrefix(ipNet, isAdd)
+		if err != nil {
+			return errors.Wrapf(err, "Couldn't configure SNAT prefix")
+		}
+	}
+	return nil
+}
+
+func (s *Server) ipamUpdateHandler(poolUpdate *proto.IPAMPoolUpdate, prevPoolUpdate *proto.IPAMPoolUpdate) (err error) {
+	if prevPoolUpdate == nil {
```

Comment:
Same this, can probably be unrolled in the caller (one if in the add, the other in the del)